File: src-tauri/plugins/cosmos/ios/Sources/CosmosPlugin/CosmosPlugin.swift
import SwiftRs
import Tauri
import UIKit

/// Argument shapes.
///
/// Flat rather than nested, because `Prompt` is `#[serde(flatten)]`-ed into the request
/// structs on the Rust side — see `src-tauri/plugins/cosmos/src/models.rs`. Non-optional
/// on every required field, so a malformed payload fails to decode rather than reaching
/// the Keychain with an empty string.
class AuthStoreArgs: Decodable {
    let key: String
    let value: String
    let title: String
    let reason: String
    let cancel: String
}

class AuthReadArgs: Decodable {
    let key: String
    let title: String
    let reason: String
    let cancel: String
}

class AuthDeleteArgs: Decodable {
    let key: String
}

class ShareArgs: Decodable {
    let text: String
    let title: String?
}

class ExcludeBackupArgs: Decodable {
    let path: String
}

/// The bridge. Decodes arguments, forwards, and turns the answer back into an `Invoke`
/// resolution — no logic of its own, which is what keeps the two capabilities behind it
/// from growing shared behaviour.
///
/// Every rejection carries its `Failure` token as the `code`, because that is the field
/// `src-tauri/plugins/cosmos/src/error.rs` reads the classification back out of. Rejecting
/// with a message alone would reach the frontend as an unclassified `failed`, which paints
/// a red error line over a user who simply tapped "cancel".
class CosmosPlugin: Plugin {

    /// Keychain and LocalAuthentication calls BLOCK while their sheet is up. Run on the
    /// main thread they would deadlock the very run loop that has to draw it, so every
    /// auth command hops onto this queue first. UIKit presentation goes the other way —
    /// see `shareText`.
    private let queue = DispatchQueue(label: "lat.cosmospay.wallet.auth", qos: .userInitiated)

    @objc public func authStatus(_ invoke: Invoke) throws {
        queue.async {
            let status = DeviceAuth.status()
            var payload: [String: Any] = [
                "available": status.available,
                "biometry": status.biometry.rawValue,
            ]
            // Omitted entirely when there is none — `AuthStatus::reason` is
            // `#[serde(default)]` on the Rust side precisely so this can stay absent
            // rather than carrying a null that means the same thing.
            if let reason = status.reason {
                payload["reason"] = reason.rawValue
            }
            invoke.resolve(payload)
        }
    }

    @objc public func authStore(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(AuthStoreArgs.self)
        queue.async {
            do {
                try DeviceAuth.store(
                    key: args.key,
                    value: args.value,
                    prompt: Prompt(title: args.title, reason: args.reason, cancel: args.cancel)
                )
                invoke.resolve()
            } catch {
                invoke.reject(error)
            }
        }
    }

    @objc public func authRead(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(AuthReadArgs.self)
        queue.async {
            do {
                let value = try DeviceAuth.read(
                    key: args.key,
                    prompt: Prompt(title: args.title, reason: args.reason, cancel: args.cancel)
                )
                invoke.resolve(["value": value])
            } catch {
                invoke.reject(error)
            }
        }
    }

    @objc public func authDelete(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(AuthDeleteArgs.self)
        queue.async {
            // Always resolves: every caller is on a cleanup path and "there was nothing to
            // delete" is success. See `desktop.rs` for the same choice.
            DeviceAuth.delete(key: args.key)
            invoke.resolve()
        }
    }

    @objc public func shareText(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(ShareArgs.self)
        DispatchQueue.main.async {
            guard let controller = self.manager.viewController else {
                // No sheet appeared. `failed` rather than `cancelled`, and the difference
                // is what makes `src/lib/share.ts` fall back to the clipboard instead of
                // assuming the user declined.
                invoke.reject(Failure.failed.rawValue, code: Failure.failed.rawValue)
                return
            }
            do {
                try ShareSheet.present(from: controller, text: args.text, title: args.title)
                invoke.resolve()
            } catch {
                invoke.reject(error.localizedDescription, code: Failure.failed.rawValue)
            }
        }
    }

    /// iOS has no hardware back button, so nothing ever emits the event that leads here.
    ///
    /// Resolves rather than refusing, and does NOT exit: an iOS app that terminates itself
    /// looks to the user like a crash, and Apple rejects it under the Human Interface
    /// Guidelines. If this is ever reached, doing nothing is the correct behaviour.
    @objc public func appExit(_ invoke: Invoke) throws {
        invoke.resolve()
    }

    /// Keep the app-data directory out of iCloud.
    ///
    /// This is the one platform where the wallet's storage is backup-eligible by default.
    /// Without it, `cosmos-wallet.json` — the AES-GCM sealed vault, and the plaintext wallet
    /// list beside it — is restored onto whatever device the user sets up next, where the
    /// password is the only thing left between an attacker and the seed, offline and at
    /// their own pace. The device-unlock envelope does not travel: its key is
    /// `.biometryCurrentSet` and stays on the phone that made it. The vault does.
    ///
    /// The DIRECTORY, not the file. `tauri-plugin-store` rewrites its file, and a resource
    /// value set on an inode does not survive that file being replaced; on a directory it
    /// covers everything written inside it, including the rewrite. The directory is created
    /// first when it does not exist yet, because this runs before the store's first write
    /// and `setResourceValues` on a missing path fails.
    ///
    /// Idempotent — setting the flag on an already-excluded directory is a no-op — so the
    /// once-per-launch call from `src/lib/storage.ts` needs no state on either side.
    @objc public func excludeFromBackup(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(ExcludeBackupArgs.self)
        queue.async {
            do {
                var url = URL(fileURLWithPath: args.path, isDirectory: true)
                if !FileManager.default.fileExists(atPath: url.path) {
                    try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
                }
                var values = URLResourceValues()
                values.isExcludedFromBackup = true
                try url.setResourceValues(values)
                invoke.resolve()
            } catch {
                // `failed`, with the platform's own sentence. Nothing here is classifiable
                // as a device-auth outcome, and the caller only logs it.
                invoke.reject(error.localizedDescription, code: Failure.failed.rawValue)
            }
        }
    }
}

/// Reject with the classification in `code`, whatever shape the error arrived in.
extension Invoke {
    fileprivate func reject(_ error: Error) {
        guard let classified = error as? DeviceAuthError else {
            // Not one of ours — a decoding fault, a UIKit throw. `failed` is the honest
            // answer: inventing a friendlier classification would be inventing a fact.
            reject(error.localizedDescription, code: Failure.failed.rawValue)
            return
        }
        reject(classified.detail ?? classified.failure.rawValue, code: classified.failure.rawValue)
    }
}

@_cdecl("init_plugin_cosmos")
func initPlugin() -> Plugin {
    return CosmosPlugin()
}

File: src-tauri/plugins/cosmos/android/src/main/java/lat/cosmospay/plugin/cosmos/CosmosPlugin.kt
package lat.cosmospay.plugin.cosmos

import android.app.Activity
import android.webkit.WebView
import androidx.activity.OnBackPressedCallback
import androidx.fragment.app.FragmentActivity
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

/**
 * Argument shapes.
 *
 * Flat rather than nested, because `Prompt` is `#[serde(flatten)]`-ed into the request
 * structs on the Rust side — see `src-tauri/plugins/cosmos/src/models.rs`. `lateinit` on
 * every required field is what turns a malformed payload into a crash in the bridge rather
 * than a silent empty string reaching the Keystore.
 */
@InvokeArg
internal class AuthStoreArgs {
    lateinit var key: String
    lateinit var value: String
    lateinit var title: String
    lateinit var reason: String
    lateinit var cancel: String
}

@InvokeArg
internal class AuthReadArgs {
    lateinit var key: String
    lateinit var title: String
    lateinit var reason: String
    lateinit var cancel: String
}

@InvokeArg
internal class AuthDeleteArgs {
    lateinit var key: String
}

@InvokeArg
internal class ShareArgs {
    lateinit var text: String
    var title: String? = null
}

/**
 * The bridge. Parses arguments, forwards, and turns the answer back into an `Invoke`
 * resolution — no logic of its own, which is what keeps the two capabilities behind it
 * from growing shared behaviour.
 *
 * Every rejection carries its `Failure` token as the `code`, because that is the field
 * `src-tauri/plugins/cosmos/src/error.rs` reads the classification back out of. Rejecting
 * with a message alone would reach the frontend as an unclassified `failed`, which paints
 * a red error line over a user who simply tapped "cancel".
 */
@TauriPlugin
class CosmosPlugin(private val activity: Activity) : Plugin(activity) {

    companion object {
        /**
         * The DOM event dispatched on `window` when the hardware back button is pressed.
         *
         * Listened for by `src/app/WalletApp.tsx` through `nativeListen`, and namespaced
         * because `window` is a shared bus. Tauri's own `trigger()` / `addPluginListener`
         * pair is deliberately NOT used: `addPluginListener` invokes
         * `plugin:cosmos|register_listener`, which Tauri implements in this base class but
         * does not forward to from Rust, so a custom plugin would have to hold the
         * `Channel` and route it itself — real plumbing on the signing path's own crate,
         * for one zero-argument notification.
         */
        const val BACK_PRESSED = "cosmos:backPressed"
    }

    /** Held from `load` so the back callback can reach the page. */
    private var webView: WebView? = null

    /**
     * The activity the biometric sheet attaches to.
     *
     * `BiometricPrompt` needs a `FragmentActivity` — it hosts itself in a fragment. Tauri's
     * own activity is one, so this cast holds; it is checked rather than forced because the
     * failure mode of `as` here would be a crash on the user's unlock screen, and
     * `unsupported` is a state the wallet already knows how to present.
     */
    private val host: FragmentActivity?
        get() = activity as? FragmentActivity

    /**
     * Take over the hardware back button.
     *
     * The wallet owns a navigation stack (`src/lib/screens.ts`) that the WebView knows
     * nothing about, so the default behaviour — finish the activity — closes the app from
     * whatever screen the user is on. Every press is forwarded to the web layer instead,
     * which pops its own stack and calls `app_exit` when there is nowhere left to go.
     *
     * `OnBackPressedDispatcher` rather than an override in `MainActivity`: that file lives
     * under `src-tauri/gen/android/`, which `tauri android init` regenerates, so anything
     * written into it is one re-init away from silently disappearing. A callback registered
     * from here travels with the plugin.
     */
    override fun load(webView: WebView) {
        super.load(webView)
        this.webView = webView
        val host = this.host ?: return
        host.onBackPressedDispatcher.addCallback(
            host,
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    // Already on the UI thread — `OnBackPressedDispatcher` dispatches there,
                    // and that is the only thread `evaluateJavascript` may be called from.
                    this@CosmosPlugin.webView?.evaluateJavascript(
                        "window.dispatchEvent(new Event('$BACK_PRESSED'))",
                        null,
                    )
                }
            },
        )
    }

    @Command
    fun authStatus(invoke: Invoke) {
        val context = host
        if (context == null) {
            // Reported as a STATUS rather than a rejection: "what can this device do" is a
            // question every build may ask, and "nothing" is a valid answer to it.
            invoke.resolve(statusObject(Status(false, Biometry.GENERIC, Failure.UNSUPPORTED)))
            return
        }
        invoke.resolve(statusObject(DeviceAuth.status(context)))
    }

    @Command
    fun authStore(invoke: Invoke) {
        val args = invoke.parseArgs(AuthStoreArgs::class.java)
        val context = host ?: return invoke.rejectWith(Failure.UNSUPPORTED, null)
        DeviceAuth.store(
            context,
            args.key,
            args.value,
            Prompt(args.title, args.reason, args.cancel),
            object : Outcome<Unit> {
                override fun ok(value: Unit) = invoke.resolve()
                override fun fail(failure: Failure, detail: String?) = invoke.rejectWith(failure, detail)
            },
        )
    }

    @Command
    fun authRead(invoke: Invoke) {
        val args = invoke.parseArgs(AuthReadArgs::class.java)
        val context = host ?: return invoke.rejectWith(Failure.UNSUPPORTED, null)
        DeviceAuth.read(
            context,
            args.key,
            Prompt(args.title, args.reason, args.cancel),
            object : Outcome<String> {
                override fun ok(value: String) = invoke.resolve(JSObject().put("value", value))
                override fun fail(failure: Failure, detail: String?) = invoke.rejectWith(failure, detail)
            },
        )
    }

    @Command
    fun authDelete(invoke: Invoke) {
        val args = invoke.parseArgs(AuthDeleteArgs::class.java)
        // Resolves even with no activity to speak of: every caller is on a cleanup path and
        // "there was nothing to delete" is success. See `desktop.rs` for the same choice.
        DeviceAuth.delete(activity, args.key)
        invoke.resolve()
    }

    @Command
    fun shareText(invoke: Invoke) {
        val args = invoke.parseArgs(ShareArgs::class.java)
        try {
            ShareSheet.share(activity, args.text, args.title)
            invoke.resolve()
        } catch (err: Throwable) {
            // No sheet appeared. `FAILED` rather than `CANCELLED`, and the difference is
            // what makes `src/lib/share.ts` fall back to the clipboard instead of assuming
            // the user declined.
            invoke.rejectWith(Failure.FAILED, err.message)
        }
    }

    /**
     * Close the app, for the one case the web layer cannot handle: back pressed with an
     * empty navigation stack.
     *
     * `finish()` rather than `exitProcess`: it lets Android tear the activity down the way
     * it does for every other app, so the task is left in a state the launcher can resume.
     */
    @Command
    fun appExit(invoke: Invoke) {
        activity.runOnUiThread { activity.finish() }
        invoke.resolve()
    }

    /**
     * Android needs no code for this, and the method exists anyway.
     *
     * Backup here is a MANIFEST decision, not a runtime one: `scripts/native-permissions.ts`
     * writes `android:allowBackup="false"` for cloud backup and the `dataExtractionRules`
     * that govern device-to-device transfer, both of which are already in force by the time
     * any of this runs. iOS has no equivalent — `NSURLIsExcludedFromBackupKey` is set on the
     * URL at runtime — so the command exists for that platform and answers honestly here.
     *
     * Resolves rather than refusing. `src/lib/storage.ts` only asks on iOS, so nothing
     * reaches this today; if something ever does, "the directory is not in a backup" is a
     * true statement on Android and an error would be a false one.
     */
    @Command
    fun excludeFromBackup(invoke: Invoke) {
        invoke.resolve()
    }

    private fun statusObject(status: Status): JSObject = JSObject()
        .put("available", status.available)
        .put("biometry", status.biometry)
        .apply {
            // Omitted entirely when there is none — `AuthStatus::reason` is
            // `#[serde(default)]` on the Rust side precisely so this can stay absent
            // rather than carrying a null that means the same thing.
            status.reason?.let { put("reason", it.token) }
        }

    private fun Invoke.rejectWith(failure: Failure, detail: String?) =
        reject(detail ?: failure.token, failure.token)
}
